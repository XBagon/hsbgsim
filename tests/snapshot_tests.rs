use hsbgsim::*;
use insta::assert_yaml_snapshot;
use rand::SeedableRng;
use rand_xoshiro::Xoshiro256PlusPlus;
use std::collections::BTreeMap;
use test_case::test_matrix;

#[test_matrix(
    0..100
)]
pub fn random_board_snapshot(seed: u64) {
    let mut game = Game::with_seed(seed);

    let mut rng = Xoshiro256PlusPlus::seed_from_u64(!seed);

    for _ in 0..7 {
        let minion = game.instantiate_minion(MinionVariant::random(&mut rng));
        game.position_minion(minion, PlayerId::Bottom).unwrap();

        let minion = game.instantiate_minion(MinionVariant::random(&mut rng));
        game.position_minion(minion, PlayerId::Top).unwrap();
    }

    game.initialize();

    let recording = game.run_and_record_events();
    let mut minions: BTreeMap<_, _> =
        game.minion_instances.iter().map(|(mi_id, minion)| (mi_id, minion.variant)).collect();

    //TODO: better replay format
    assert_yaml_snapshot!(format!("random_board_snapshot_{}", seed), (minions, recording));
}
