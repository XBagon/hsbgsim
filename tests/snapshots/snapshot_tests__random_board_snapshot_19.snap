---
source: tests/snapshot_tests.rs
expression: "(minions, recording)"
---
- ? idx: 1
    version: 1
  : TitusRivendare
  ? idx: 2
    version: 1
  : Niuzao
  ? idx: 3
    version: 1
  : SoulJuggler
  ? idx: 4
    version: 1
  : WitheredSpearhide
  ? idx: 5
    version: 1
  : BrannBronzebeard
  ? idx: 6
    version: 1
  : ObsidianRavager
  ? idx: 7
    version: 1
  : XyloBones
  ? idx: 8
    version: 1
  : PrimalfinLookout
  ? idx: 9
    version: 1
  : SlyRaptor
  ? idx: 10
    version: 1
  : DeepBlueCrooner
  ? idx: 11
    version: 1
  : TitusRivendare
  ? idx: 12
    version: 1
  : BrinyBootlegger
  ? idx: 13
    version: 1
  : FacelessDisciple
  ? idx: 14
    version: 1
  : UpbeatFlutist
- - ProposeAttack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 5
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 5
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 5
        version: 1
      amount: 6
      source:
        idx: 2
        version: 1
  - TakeDamage:
      target:
        idx: 2
        version: 1
      amount: 2
      source:
        idx: 5
        version: 1
  - DeathCheck: ~
  - Death:
      minion:
        idx: 5
        version: 1
  - AfterAttack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 5
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 1
        version: 1
      defender:
        idx: 14
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 1
        version: 1
      defender:
        idx: 14
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 14
        version: 1
      amount: 1
      source:
        idx: 1
        version: 1
  - TakeDamage:
      target:
        idx: 1
        version: 1
      amount: 2
      source:
        idx: 14
        version: 1
  - DeathCheck: ~
  - AfterAttack:
      attacker:
        idx: 1
        version: 1
      defender:
        idx: 14
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 4
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 4
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 13
        version: 1
      amount: 2
      source:
        idx: 4
        version: 1
  - TakeDamage:
      target:
        idx: 4
        version: 1
      amount: 6
      source:
        idx: 13
        version: 1
  - DeathCheck: ~
  - Death:
      minion:
        idx: 4
        version: 1
  - AfterAttack:
      attacker:
        idx: 4
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 12
        version: 1
      amount: 3
      source:
        idx: 3
        version: 1
  - TakeDamage:
      target:
        idx: 3
        version: 1
      amount: 4
      source:
        idx: 12
        version: 1
  - DeathCheck: ~
  - AfterAttack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 6
        version: 1
      defender:
        idx: 7
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 6
        version: 1
      defender:
        idx: 7
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 7
        version: 1
      amount: 7
      source:
        idx: 6
        version: 1
  - TakeDamage:
      target:
        idx: 6
        version: 1
      amount: 6
      source:
        idx: 7
        version: 1
  - DeathCheck: ~
  - Death:
      minion:
        idx: 7
        version: 1
  - AfterAttack:
      attacker:
        idx: 6
        version: 1
      defender:
        idx: 7
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 2
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 2
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 2
        version: 1
      amount: 3
      source:
        idx: 3
        version: 1
  - TakeDamage:
      target:
        idx: 3
        version: 1
      amount: 6
      source:
        idx: 2
        version: 1
  - DeathCheck: ~
  - Death:
      minion:
        idx: 2
        version: 1
  - Death:
      minion:
        idx: 3
        version: 1
  - AfterAttack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 2
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 6
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 6
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 13
        version: 1
      amount: 7
      source:
        idx: 6
        version: 1
  - TakeDamage:
      target:
        idx: 6
        version: 1
      amount: 6
      source:
        idx: 13
        version: 1
  - DeathCheck: ~
  - Death:
      minion:
        idx: 6
        version: 1
  - Death:
      minion:
        idx: 13
        version: 1
  - AfterAttack:
      attacker:
        idx: 6
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 9
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 9
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 12
        version: 1
      amount: 1
      source:
        idx: 9
        version: 1
  - TakeDamage:
      target:
        idx: 9
        version: 1
      amount: 4
      source:
        idx: 12
        version: 1
  - DeathCheck: ~
  - Death:
      minion:
        idx: 12
        version: 1
  - Death:
      minion:
        idx: 9
        version: 1
  - AfterAttack:
      attacker:
        idx: 9
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 8
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 8
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 1
        version: 1
      amount: 3
      source:
        idx: 8
        version: 1
  - TakeDamage:
      target:
        idx: 8
        version: 1
      amount: 1
      source:
        idx: 1
        version: 1
  - DeathCheck: ~
  - AfterAttack:
      attacker:
        idx: 8
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 10
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 10
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 10
        version: 1
      amount: 1
      source:
        idx: 11
        version: 1
  - TakeDamage:
      target:
        idx: 11
        version: 1
      amount: 2
      source:
        idx: 10
        version: 1
  - DeathCheck: ~
  - AfterAttack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 10
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 10
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 10
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 1
        version: 1
      amount: 2
      source:
        idx: 10
        version: 1
  - TakeDamage:
      target:
        idx: 10
        version: 1
      amount: 1
      source:
        idx: 1
        version: 1
  - DeathCheck: ~
  - Death:
      minion:
        idx: 10
        version: 1
  - Death:
      minion:
        idx: 1
        version: 1
  - AfterAttack:
      attacker:
        idx: 10
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 14
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 14
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 14
        version: 1
      amount: 1
      source:
        idx: 11
        version: 1
  - TakeDamage:
      target:
        idx: 11
        version: 1
      amount: 2
      source:
        idx: 14
        version: 1
  - DeathCheck: ~
  - AfterAttack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 14
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 14
        version: 1
      defender:
        idx: 11
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 14
        version: 1
      defender:
        idx: 11
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 11
        version: 1
      amount: 2
      source:
        idx: 14
        version: 1
  - TakeDamage:
      target:
        idx: 14
        version: 1
      amount: 1
      source:
        idx: 11
        version: 1
  - DeathCheck: ~
  - AfterAttack:
      attacker:
        idx: 14
        version: 1
      defender:
        idx: 11
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 8
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 8
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 8
        version: 1
      amount: 1
      source:
        idx: 11
        version: 1
  - TakeDamage:
      target:
        idx: 11
        version: 1
      amount: 3
      source:
        idx: 8
        version: 1
  - DeathCheck: ~
  - Death:
      minion:
        idx: 8
        version: 1
  - Death:
      minion:
        idx: 11
        version: 1
  - AfterAttack:
      attacker:
        idx: 11
        version: 1
      defender:
        idx: 8
        version: 1
      is_outer_phase: true
  - DeathCheck: ~
  - End: TopWin

